specification/under_development/Instruction_Set.md

```diff
@@ -0,0 +1,52 @@
+
+## Quantum Instruction Set (QIS)
+
+The QIR specification contains a [table](../QuantumInstructions.md) of commonly used instructions along with their signatures and a description of their functionality.
+Backends are **not** required to support all of these. Instead, each backend will declare which of these instructions it supports. We encourage to make instructions that are supported by a context independent implementation in terms of other instructions as a library rather than listing them as part of the backend specification. A library provided either in the form of a bitcode file can be linked in as part of a QIR compilation stage. 
+
+
+All quantum instructions are represented by LLVM external functions.
+Quantum instructions may take qubits, doubles, or integers as parameters,
+and should all have no return; that is, they should be void.
+
+Measurements should take an offset into the `ClassicalStorage` global as a parameter.
+The measurement result should be stored into the corresponding bit in `ClassicalStorage`.
+
+The LLVM functions that implement the quantum instruction set should all have
+names that start with `__quantum__qis__`.
+
+QIR does not specify the contents of the quantum instruction set.
+However, in order to ensure some amount of uniformity, implementations that provide
+any of the following quantum instructions must match the specified definition:
+
+| Operation Name | LLVM Function Declaration  | Description | Matrix |
+|----------------|----------------------------|-------------|--------|
+| CCx, CCNOT, Toffoli | `__quantum__qis__toffoli__body (%Qubit control1, %Qubit control1, %Qubit target)` | Toffoli or doubly-controlled X | ![latex](https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+%5Cbegin%7Bbmatrix%7D+1+%26+0+%26+0+%26+0+%26+0+%26+0+%26+0+%26+0+%5C%5C+0+%26+1+%26+0+%26+0+%26+0+%26+0+%26+0+%26+0+%5C%5C+0+%26+0+%26+1+%26+0+%26+0+%26+0+%26+0+%26+0+%5C%5C+0+%26+0+%26+0+%26+1+%26+0+%26+0+%26+0+%26+0+%5C%5C+0+%26+0+%26+0+%26+0+%26+1+%26+0+%26+0+%26+0+%5C%5C+0+%26+0+%26+0+%26+0+%26+0+%26+1+%26+0+%26+0+%5C%5C+0+%26+0+%26+0+%26+0+%26+0+%26+0+%26+0+%26+1+%5C%5C+0+%26+0+%26+0+%26+0+%26+0+%26+0+%26+1+%26+0+%5C%5C+%5Cend%7Bbmatrix%7D) |
+| Cx, CNOT | `__quantum__qis__cnot__body (%Qubit control, %Qubit target)` | CNOT or singly-controlled X | ![latex](https://render.githubusercontent.com/render/math?math=%5Cdisplaystyle+%5Cbegin%7Bbmatrix%7D+1+%26+0+%26+0+%26+0+%5C%5C+0+%26+1+%26+0+%26+0+%5C%5C+0+%26+0+%26+0+%26+1+%5C%5C+0+%26+0+%26+1+%26+0+%5C%5C+%5Cend%7Bbmatrix%7D) |
```

Comment:
These signatures pass qubits by value instead of as pointers, which doesn't match the spec patterns. They should use `%Qubit*` throghout.

Reply:
Ups - I copy-pasted that table and didn't notice... Thanks!